python/python/lance/util.py

```diff
@@ -40,3 +45,118 @@ def sanitize_ts(ts: ts_types) -> datetime:
+    """KMean model for clustering.
+
+    It works with 2-D arrays of float32 type,
+    and support distance metrics: "l2", "cosine", "dot".
+
+    Note, you must train the kmeans model by calling :meth:`fit` before
+    calling :meth:`predict`.
+    Calling `fit()` again will reset the model.
+
+    Currently, the initial centroids are initialized randomly. ``kmean++``
+    is implemented but not exposed yet.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> import lance
+    >>> data = np.random.randn(1000, 128).astype(np.float32)
+    >>> kmeans = lance.util.KMeans(8, metric_type="cosine")
+    >>> kmeans.fit(data)
+    >>> centroids = np.stack(kmeans.centroids.to_numpy(zero_copy_only=False))
+    >>> clusters = kmeans.predict(data)
+    """
+
+    def __init__(
+        self,
+        k: int,
+        metric_type: Literal["l2", "dot", "cosine"] = "l2",
+        max_iters: int = 50,
+    ):
+        """Create a KMeans model.
+
+        Parameters
+        ----------
+        k: int
+            The number of clusters to create.
+        metric_type: str, default="l2"
+            The metric to use for calculating distances between vectors.
+            Supported distance metrics: "l2", "cosine", "dot"
+        max_iters: int
+            The maximum number of iterations to run the KMeans algorithm. Default: 50.
+        """
+        metric_type = metric_type.lower()
+        if metric_type not in ["l2", "dot", "cosine"]:
+            raise ValueError(
+                f"metric_type must be one of 'l2', 'dot', 'cosine', got: {metric_type}"
+            )
+        self.k = k
+        self._metric_type = metric_type
+        self._kmeans = _KMeans(k, metric_type, max_iters=max_iters)
+
+    def __repr__(self) -> str:
+        return f"lance.KMeans(k={self.k}, metric_type={self._metric_type})"
+
+    @property
+    def centroids(self) -> Optional[pa.FixedShapeTensorArray]:
+        """Returns the centroids of the model,
+
+        Returns None if the model is not trained.
+        """
+        ret = self._kmeans.centroids()
+        if ret is None:
+            return None
+        shape = (ret.type.list_size,)
+        tensor_type = pa.fixed_shape_tensor(ret.type.value_type, shape)
+        return pa.FixedShapeTensorArray.from_storage(tensor_type, ret)
+
+    def _to_fixed_size_list(self, data: pa.Array) -> pa.FixedSizeListArray:
+        if isinstance(data, pa.FixedSizeListArray):
+            if data.type.value_type != pa.float32():
+                raise ValueError(
+                    f"Array must be float32 type, got: {data.type.value_type}"
+                )
+            return data
+        elif isinstance(data, pa.FixedShapeTensorArray):
+            if len(data.type.shape) != 1:
+                raise ValueError(
+                    f"Fixed shape tensor array must be a 1-D array, "
+                    f"got {len(data.type.shape)}-D"
+                )
+            return self._to_fixed_size_list(data.storage)
```

Comment:
maybe to avoid the PyArrow upgrade we could instead just write:

```suggestion
        elif (isinstance(data, pa.ExtensionArray)
              and isinstance(data.storage, pa.FixedSizeListArray)):
            return self._to_fixed_size_list(data.storage)
```

Reply:
fixed.